### PR TITLE
[BAEL-6548] Determine empty row in Microsoft Excel

### PR DESCRIPTION
**Pull request by sdhiray7**
Added the code and tests for checking if the row is empty using the following libraries:

1. Apache POI
2. JExcel
3. Fastexcel

Test cases for each of the library with a file containing a few records and an empty file.

### CODE REVIEW

File: apache-poi/src/test/java/com/baeldung/fastexcel/FastexcelIntegrationTest.java [hunk 1]
```diff
@@ -38,7 +43,18 @@ public void whenParsingExcelFile_thenCorrect() throws IOException {
...
         assertEquals("20", data.get(3).get(1));
     }
-
+    
+    @Test
+    public void whenParsingFastExcelFile_thenDetectEmptyRow() throws IOException {
+    	try (FileInputStream file = new FileInputStream(fileLocation); ReadableWorkbook wb = new ReadableWorkbook(file)) {
+            Sheet sheet = wb.getFirstSheet();
+            try (Stream<Row> rows = sheet.openStream()) {
+            	assertEquals(false, fastexcelHelper.isRowEmpty(rows.reduce((first, second) -> second).get()));
```
Comment (ashleyfrieze) [inline apache-poi/src/test/java/com/baeldung/fastexcel/FastexcelIntegrationTest.java:52]: `assertFalse` ?

Comment (ashleyfrieze) [reply apache-poi/src/test/java/com/baeldung/fastexcel/FastexcelIntegrationTest.java:52]: This `reduce` statement is a bit hard to understand

Comment (sdhiray7) [reply apache-poi/src/test/java/com/baeldung/fastexcel/FastexcelIntegrationTest.java]: I have changed it as per the review comments.

File: apache-poi-3/src/main/java/com/baeldung/fastexcel/FastexcelHelper.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
+package com.baeldung.fastexcel;
+
+import org.dhatim.fastexcel.reader.Cell;
+import org.dhatim.fastexcel.reader.Row;
+
+public class FastexcelHelper {
+	public boolean isRowEmpty(Row row) {
```
Comment (ashleyfrieze) [inline apache-poi-3/src/main/java/com/baeldung/fastexcel/FastexcelHelper.java:7]: This is not formatted as per the standard. You're using tabs, where we should be using spaces.

File: apache-poi-3/src/main/java/com/baeldung/jexcel/JExcelHelper.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+import java.io.File;
+import java.io.IOException;
+import jxl.write.*;
+import jxl.write.Number;
+import jxl.format.Colour;
+
+public class JExcelHelper {
+    
+    public boolean isRowEmpty(Cell[] row) {
+    	if (row == null) {
```
Comment (ashleyfrieze) [inline apache-poi-3/src/main/java/com/baeldung/jexcel/JExcelHelper.java:19]: This is using a mixture of tabs and spaces

File: apache-poi-3/src/test/java/com/baeldung/jexcel/JExcelDetectEmptyRowUnitTest.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+import org.junit.Before;
+import org.junit.Test;
+
+import jxl.Cell;
+import jxl.Sheet;
+import jxl.Workbook;
+import jxl.read.biff.BiffException;
+
+
+public class JExcelDetectEmptyRowUnitTest {
```
Comment (ashleyfrieze) [inline apache-poi-3/src/test/java/com/baeldung/jexcel/JExcelDetectEmptyRowUnitTest.java:17]: Needs to be formatted as per the standard - this uses tabs

File: apache-poi-3/src/test/java/com/baeldung/jexcel/JExcelDetectEmptyRowUnitTest.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+	private JExcelHelper jexcelHelper;
+	private static String FILE_NAME = "consumer_info_with_empty_row.xls";
+	private String fileLocation;
+
+	@Before
+	public void loadExcelFile() throws IOException {
+
+		fileLocation = "src/main/resources/" + FILE_NAME;
```
Comment (ashleyfrieze) [inline apache-poi-3/src/test/java/com/baeldung/jexcel/JExcelDetectEmptyRowUnitTest.java:26]: This does not need to be initialized this way. `fileLocation` always has the same value and so should be `private static final String FILE_LOCATION` initialized in the class' static initializer list

File: apache-poi-3/src/test/java/com/baeldung/poi/columnnames/ExcelUtilsUnitTest.java [hunk 6]
```diff
@@ -8,37 +8,72 @@
...
-    workbook.close();
-}
+	private static final String XLS_TEST_FILE_PATH = "src/main/resources/consumer_info.xls";
+	private static final String XLSX_TEST_FILE_PATH = "src/main/resources/food_info.xlsx";
+	private static final String XLS_EMPTY_ROW_FILE_PATH = "src/main/resources/consumer_info_with_empty_row.xls";
+	private static final String XLSX_EMPTY_ROW_FILE_PATH = "src/main/resources/consumer_info_with_empty_row.xlsx";
+	private static final String SHEET_NAME = "Sheet1";
+	
+	@Test
+	public void givenExcelFileWithXLSXFormat_whenGetColumnNames_thenReturnsColumnNames() throws IOException {
```
Comment (ashleyfrieze) [inline apache-poi-3/src/test/java/com/baeldung/poi/columnnames/ExcelUtilsUnitTest.java:23]: reformat with spaces

**Review by ashleyfrieze — CHANGES_REQUESTED** [review]

File: apache-poi-3/src/test/java/com/baeldung/fastexcel/FastexcelDetectEmptyRowUnitTest.java [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    public void loadExcelFile() throws IOException {
+        fileLocation = "src/main/resources/" + FILE_NAME;
+        fastexcelHelper = new FastexcelHelper();
+    }
+
+    @Test
+    public void whenParsingFastExcelFile_thenDetectEmptyRow() throws IOException {
+        try (FileInputStream file = new FileInputStream(fileLocation); ReadableWorkbook wb = new ReadableWorkbook(file)) {
+            Sheet sheet = wb.getFirstSheet();
+            try (Stream<Row> rows = sheet.openStream()) {
```
Comment (ashleyfrieze) [inline apache-poi-3/src/test/java/com/baeldung/fastexcel/FastexcelDetectEmptyRowUnitTest.java:34]: Why is this implementation not possible?

```java
// assert all rows are empty
try (Stream<Row> rows = sheet.openStream()) {
    boolean isEmpty = rows.allMatch(fastexcelHelper::isRowEmpty);

    assertTrue(isEmpty);
}

// assert last row is empty
try (Stream<Row> rows = sheet.openStream()) {
    Optional<Row> last = rows.reduce((first, second) -> second);
    boolean isLastRowEmpty = last.map(fastexcelHelper::isRowEmpty).orElseThrow();
    assertTrue(isLastRowEmpty);
}
```

Comment (sdhiray7) [reply apache-poi-3/src/test/java/com/baeldung/fastexcel/FastexcelDetectEmptyRowUnitTest.java:34]: Done, I added test cases for an empty file and a file with a few records but empty rows.

File: apache-poi-3/src/test/java/com/baeldung/jexcel/JExcelDetectEmptyRowUnitTest.java (hunk 4 above)
Comment (sdhiray7) [reply apache-poi-3/src/test/java/com/baeldung/jexcel/JExcelDetectEmptyRowUnitTest.java:17]: Done, please check.

**Pull request merged**